cache/pmdk_cache.cc

```diff
@@ -0,0 +1,438 @@
+  // insertion into persistent tier
+  const Slice& unpacked_val = unpack(value);
+  // TODO: better estimation of persistent total charge:
+
+  size_t persistent_charge = sizeof(PersistentEntry) + 
+                              key.size() + 
+                              unpacked_val.size() + 
+                              // rough estimation of hash map entry: cache line size,
+                              // since each hash map node requires its mutex padded
+                              // to cache line.
+                              _SC_LEVEL2_CACHE_LINESIZE;
+
+  po::transaction::run(pop_, [&, unpacked_val, persistent_charge, s_p, deleter, pack] {
+    bool fake_insertion = false;
+    {
+      MutexLock l(&mutex_);
+      EvictFromLRU(persistent_charge, &last_reference_list);
+      // TODO: calculate charge and refuse insert if cache is full
+      if ((usage_ + persistent_charge) > persistent_capacity_ && handle == nullptr){
+        if (handle == nullptr){
+          // Don't insert the entry but still return ok, as if the entry inserted
+          // into cache and get evicted immediately.
+          fake_insertion = true;
+        } else {
+          *handle == nullptr;
+          *s_p = Status::Incomplete("Insert failed due to LRU cache being full.");
+        }
+      } else {
+        // create new persistent entry
+        auto p_entry = po::make_persistent<PersistentEntry>();
+        p_entry->persist_charge = persistent_charge;
+        p_entry->key_size = key.size();
+        p_entry->key = po::make_persistent<char[]>(key.size());
+        p_entry->val_size = unpacked_val.size();
+        p_entry->val = po::make_persistent<char[]>(unpacked_val.size());
+        // memcpy from transient to persistent tier
+        pop_.memcpy_persist(p_entry->key.get(), key.data(), key.size());
+        pop_.memcpy_persist(p_entry->val.get(), unpacked_val.data(), unpacked_val.size());
+        // insert persistent entry into hash table.
+        po::persistent_ptr<PersistentEntry> old = persistent_hashtable_->Insert(hash, key, p_entry);
+        usage_ += persistent_charge;
+        if (old != nullptr){
+          *s_p = Status::OkOverwritten();
+          assert(old->InCache());
+          old->SetInCache(false);
+          if (!old->HasRefs()){
+            LRU_Remove(old);
+            size_t old_persist_charge = old->persist_charge;
+            assert(usage_ >= old_persist_charge);
+            usage_ -= old_persist_charge;
+            last_reference_list.push_back(old);
+          }
+        }
+        if (handle == nullptr){
+          LRU_Insert(p_entry);
+        } else {
+          // TODO: don't do this when insertion into transient tier is successful.
+          TransientHandle* e = GetTransientHandle(p_entry, pack);
+          e->Ref();
+          *handle = reinterpret_cast<Cache::Handle*>(e);
+        }
+      }
+    }
+    if (fake_insertion){
+      (*deleter)(key, value);
```

Comment:
The decision on whether to call `(*deleter)(key, value)` should be made based on the result of insertion into volatile tier instead of persistent tier, right?

Reply:
Yes. I think I need to rework some related logic here. See my other replies.

Reply:
Updated.